### ISSUE TITLE
Pasing options to Solvers / QuantifierEliminators / Interpolator

### ISSUE THREAD

**Issue by mikand** [pysmt/pysmt#105]
We should design a clean and simple way of passing options to the Solvers / QuantifierEliminators / Interpolator instances through the factory class

**Comment by marcogario** [comment]
This has been partially addressed by #250 . What is missing is a way of passing solver specific options.
E.g., write

```
s = Solver(name="msat", incremental=True, solver_options={"theory.arr.enabled": "false"})
```

while the option ```incremental``` is mapped to the corresponding option of each solver, there might be options that are rarely used or that are really so specific that it does not make sense to handle them using keywords argument. 

This can be implemented by providing ```SolverOptions``` with a basic ```solver_options``` case, that is then managed in a custom way by each solver.

Open points remain:

* How to decide whether an option can be promoted to a keyword option? 
* How to handle incorrect options being passed to the solver?
* What are meaningful examples in which we might want to do this? Ideally, we should add one example per solver.

**Comment by marcogario** [comment]
These options are defined in the SMT-LIB Standard. These should be supported at least by the SMT-LIB solver. Additionally, the ones marked with * are good candidates for the generic Solver interface. 

Required:

 :diagnostic-output-channel
 :regular-output-channel
 :print-success
 
Optional
 :random-seed *
 :produce-models (enables get-model) *
 :verbosity *
 :global-declarations 
 :produce-assertions (enables get-assertions)
 :produce-assignments (enables get-assignments)
 :produce-proofs (enables get-proofs)
 :produce-unsat-assumptions (enables get-unsat-assumptions)
 :produce-unsat-cores (enables get-unsat-cores)
 :reproducible-resource-limit

Note: produce-models is currently called 'generate_models". verbosity should rely on the option diagnostic-output-channel (probably to be renamed) in order to provide output on a file.

- [ ] Implement all options in the SMT-LIB wrapper
- [ ] Add random_seed and verbosity in SolverOptions
- [ ] Map random_seed and verbosity to each solver options.

**Comment by marcogario** [comment]
See http://docs.scipy.org/doc/scipy/reference/optimize.minimize-neldermead.html for reference on splitting "global" and "local" options.

**Comment by marcogario** [comment]
Completed without considering 'verbosity' in branch i105/options. Waiting for CI to open PR.

**Comment by marcogario** [comment]
Partially addressed in #338. Left for future work:

* Options for UnsatCore/Interpolator
* Add option to handle debugging file (e.g., trace API)
* Decide whether verbosity is a meaningful option (probably depends on the point above).
